objects/src/accounts/component/template/storage_entry/template_key.rs

```diff
@@ -0,0 +1,207 @@
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TemplateKey {
+    key: String,
+}
+
+impl TemplateKey {
+    /// Creates a new [TemplateKey] from the provided string.
+    ///
+    /// A [TemplateKey] serves as an identifier for storage values that are determined at
+    /// instantiation time of an [AccountComponentTemplate](super::super::AccountComponentTemplate).
+    ///
+    /// The key can consist of one or more segments separated by dots (`.`).  
+    /// Each segment must be non-empty and may contain only alphanumeric characters, underscores
+    /// (`_`), or hyphens (`-`).
+    ///
+    /// # Errors
+    ///
+    /// This method returns an error if:
+    /// - Any segment (or the whole key) is empty.
+    /// - Any segment contains invalid characters.
+    pub fn new(key: impl Into<String>) -> Result<Self, TemplateKeyError> {
+        let key: String = key.into();
+        Self::validate(&key)?;
+        Ok(Self { key })
+    }
+
+    /// Returns the key name
+    pub fn inner(&self) -> &str {
+        &self.key
+    }
+
+    /// Checks if the given string is a valid key.
+    /// A template key is valid if it's made of one or more segments that are non-empty alphanumeric
+    /// strings.
+    fn validate(key: &str) -> Result<(), TemplateKeyError> {
+        if key.is_empty() {
+            return Err(TemplateKeyError::EmptyKey);
+        }
+
+        for segment in key.split('.') {
+            if segment.is_empty() {
+                return Err(TemplateKeyError::EmptyKey);
+            }
+
+            for c in segment.chars() {
+                if !(c.is_ascii_alphanumeric() || c == '_' || c == '-') {
+                    return Err(TemplateKeyError::InvalidChar(c));
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl TryFrom<&str> for TemplateKey {
+    type Error = TemplateKeyError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        if value.starts_with("{{") && value.ends_with("}}") {
+            let inner = &value[2..value.len() - 2];
+            Self::validate(inner)?;
+
+            Ok(TemplateKey { key: inner.to_string() })
+        } else {
+            Err(TemplateKeyError::FormatError)
+        }
+    }
+}
+
+impl TryFrom<&String> for TemplateKey {
+    type Error = TemplateKeyError;
+
+    fn try_from(value: &String) -> Result<Self, Self::Error> {
+        Self::try_from(value.as_str())
+    }
+}
+
+impl core::fmt::Display for TemplateKey {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{{{{{}}}}}", self.key)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum TemplateKeyError {
+    #[error("key segment cannot be empty")]
+    EmptyKey,
+    #[error("expected string in {{...}} format")]
+    FormatError,
+    #[error(
+        "invalid character ({0}) found in TOML key (must be alphanumeric, underscore, or hyphen)"
+    )]
+    InvalidChar(char),
```

Comment:
I would not mention TOML here because we may have other ways to instantiate template keys.

Also, would it make sense to include the actual key in the error? (also relevant for `FormatError` variant).

Reply:
Yeah, I got the validation rules from TOML which is why I wrote this. Changed and added the keys.